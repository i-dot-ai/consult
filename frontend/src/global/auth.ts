export function getToken(cookie: string): string | null {
    const match = cookie.match(/authToken=([^;]+)/);
    return match ? decodeURIComponent(match[1]) : null;
}