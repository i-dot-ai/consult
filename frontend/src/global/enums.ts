export enum Routes {
    Home = "/",
    Consultations = "/consultations",
    Support = "/support",
    SignOut = "/sign-out",
}

export const getConsultationDetailUrl = (consultationId: string) => {
    return `/${Routes.Consultations}/${consultationId}`;
};

export const getResponseDetailUrl = (consultationId: string, responseId: string) => {
    return `/${Routes.Consultations}/${consultationId}/responses/${responseId}`;
};