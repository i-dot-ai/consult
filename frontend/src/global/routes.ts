export enum Routes {
  Home = "/",
  Evaluations = "/evaluations",
  Consultations = "/consultations",
  Support = "/support",
  HowItWorks = "/how-it-works",
  DataSharing = "/data-sharing",
  GetInvolved = "/get-involved",
  SignIn = "/sign-in",
  SignOut = "/sign-out",
  ApiMagicLink = "/api/magic-link/",
  ApiConsultations = "/api/consultations",
  ApiUser = "/api/user/",
  ApiAstroSignIn = "/api/astro/sign-in",
  Design = "/design",
  SupportImport = "/support/consultations/import-consultation",
  SupportUsers = "/support/users",
  SupportConsultations = "/support/consultations",
  SupportSignOff = "/support/consultations/sign-off",
  SupportThemefinder = "/support/consultations/themefinder",
}
export const getConsultationDetailUrl = (consultationId: string) => {
  return `${Routes.Consultations}/${consultationId}`;
};
export const getConsultationAnalysisUrl = (consultationId: string) => {
  return `${Routes.Consultations}/${consultationId}/analysis/`;
};

export const getThemeSignoffUrl = (consultationId: string) => {
  return `${Routes.Consultations}/${consultationId}/theme-signoff`;
};
export const getThemeSignoffDetailUrl = (consultationId: string, questionId: string) => {
  return `${Routes.Consultations}/${consultationId}/theme-signoff/${questionId}`;
};

export const getConsultationEvalUrl = (consultationId: string) => {
  return `${Routes.Evaluations}/${consultationId}/questions/`;
};

export const getQuestionDetailUrl = (
  consultationId: string,
  questionId: string,
) => {
  if (!consultationId || !questionId) {
    return "#";
  }
  return `${Routes.Consultations}/${consultationId}/responses/${questionId}`;
};

export const getRespondentDetailUrl = (
  consultationId: string,
  respondentId: string,
) => {
  if (!consultationId || !respondentId) {
    return "#";
  }
  return `${Routes.Consultations}/${consultationId}/respondent/${respondentId}`;
};
export const getApiConsultationUrl = (consultationId: string) => {
  return `${Routes.ApiConsultations}/${consultationId}`;
};
export const getApiAnswersUrl = (consultationId: string) => {
  return `/api/consultations/${consultationId}/responses/`;
};
export const getApiQuestionsUrl = (consultationId: string) => {
  return `${Routes.ApiConsultations}/${consultationId}/questions/`;
};
export const getApiQuestionUrl = (consultationId: string, questionId: string) => {
  return `${Routes.ApiConsultations}/${consultationId}/questions/${questionId}`;
};
export const getApiAnswerUrl = (
  consultationId: string,
  questionId: string,
  answerId: string,
) => {
  return `${Routes.ApiConsultations}/${consultationId}/responses/${answerId}/`;
};
export const getApiAnswerFlagUrl = (
  consultationId: string,
  questionId: string,
  answerId: string,
) => {
  return `${Routes.ApiConsultations}/${consultationId}/responses/${answerId}/toggle-flag/`;
};
export const getApiConsultationRespondentsUrl = (consultationId: string) => {
  return `${Routes.ApiConsultations}/${consultationId}/respondents/`;
};
export const getApiConsultationRespondentUrl = (
  consultationId: string,
  respondentId: string,
) => {
  return `${Routes.ApiConsultations}/${consultationId}/respondents/${respondentId}/`;
};
export const getQuestionsByRespondentUrl = (
  consultationId: string,
  respondentId: string,
) => {
  return `${Routes.ApiConsultations}/${consultationId}/questions/?respondent_id=${respondentId}`;
};

export const getApiGetSelectedThemesUrl = (consultationId: string, questionId: string) => {
  return `${Routes.ApiConsultations}/${consultationId}/questions/${questionId}/selected-themes/`;
}
export const getApiDeleteSelectedThemeUrl = (consultationId: string, questionId: string, themeId: string) => {
  return `${Routes.ApiConsultations}/${consultationId}/questions/${questionId}/selected-themes/${themeId}/`;
}
export const getApiUpdateSelectedThemeUrl = (consultationId: string, questionId: string, themeId: string) => {
  return `${Routes.ApiConsultations}/${consultationId}/questions/${questionId}/selected-themes/${themeId}/`;
}
export const getApiCreateSelectedThemeUrl = (consultationId: string, questionId: string) => {
  return `${Routes.ApiConsultations}/${consultationId}/questions/${questionId}/selected-themes/`;
}
export const getApiGetGeneratedThemesUrl = (consultationId: string, questionId: string) => {
  return `${Routes.ApiConsultations}/${consultationId}/questions/${questionId}/candidate-themes/`;
}
export const getApiSelectGeneratedThemeUrl = (consultationId: string, questionId: string, themeId: string) => {
  return `${Routes.ApiConsultations}/${consultationId}/questions/${questionId}/candidate-themes/${themeId}/select/`;
}
export const getApiConfirmSignoffUrl = (consultationId: string, questionId: string) => {
  return `${Routes.ApiConsultations}/${consultationId}/questions/${questionId}/`;
}