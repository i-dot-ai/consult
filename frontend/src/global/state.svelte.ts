function createThemeFiltersState() {
    const MAX_THEME_FILTERS = Infinity;

    let themeFilters: string[] = $state([]);

    return {
        get filters() { return themeFilters },

        applied: () => {
            return themeFilters.length > 0;
        },

        reset: () => {
            themeFilters = [];
        },

        update: (newFilter: string) => {
            if (!newFilter) {
                // Clear filters if newFilter is falsy
                themeFilters = [];
                return;
            }
            if (themeFilters.includes(newFilter)) {
                themeFilters = [...themeFilters.filter(filter => filter !== newFilter)];
            } else {
                if (themeFilters.length === MAX_THEME_FILTERS) {
                    themeFilters = [...themeFilters.slice(1), newFilter];
                } else {
                    themeFilters = [...themeFilters, newFilter];
                }
            }
        }
    }
}
export const themeFilters = createThemeFiltersState();


function createDemoFiltersState() {
    let demoFilters: {[key: string]: string[]} = $state({});

    return {
        get filters() { return demoFilters },

        reset: () => {
            demoFilters = {};
        },

        applied: (): boolean => {
            for (const key of Object.keys(demoFilters)) {
                const filterArr = demoFilters[key];

                // filterArr can be undefined or empty array
                if (filterArr && filterArr.filter(Boolean).length > 0) {
                    return true;
                }
            }
            return false;
        },

        update: (newFilterKey: string, newFilterValue: string) => {
            if (!newFilterKey || !newFilterValue) {
                // Clear filters if nothing is passed
                demoFilters = {};
                return;
            }

            const existingFilters = demoFilters[newFilterKey] || [];

            let resultFilters;
            if (existingFilters.includes(newFilterValue)) {
                // Remove filter if already added
                resultFilters = existingFilters.filter(filter => newFilterValue !== filter);
            } else {
                // Avoid duplicates when adding filters
                resultFilters = [...new Set([...existingFilters, newFilterValue])];
            }

            demoFilters = {
                ...demoFilters,
                [newFilterKey]: resultFilters,
            }
        }
    }
}
export const demoFilters = createDemoFiltersState();


function createMultiAnswerFiltersState() {
    let multiAnswerFilters: string[] = $state([]);

    return {
        get filters() { return multiAnswerFilters },

        reset: () => {
            multiAnswerFilters = [];
        },

        applied: (answer?: string): boolean => {
            // if no answer specified, return true if any applied
            if (!answer) {
                return multiAnswerFilters.length > 0;
            }
            // if answer specified, return true if it's applied
            return multiAnswerFilters.includes(answer);
        },

        update: (newFilter?: string) => {
            if (!newFilter) {
                return;
            }
            if (multiAnswerFilters.includes(newFilter)) {
                multiAnswerFilters = multiAnswerFilters.filter(filter => filter != newFilter);
            } else {
                multiAnswerFilters = [...multiAnswerFilters, newFilter];
            }
        }
    }
}
export const multiAnswerFilters = createMultiAnswerFiltersState();