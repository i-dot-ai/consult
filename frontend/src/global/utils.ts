export const getEnv = (url: string): string => {
    if (url.includes("consult-dev.ai.cabinetoffice.gov.uk")) {
        return "dev";
    }
    if (url.includes("consult.ai.cabinetoffice.gov.uk")) {
        return "prod";
    }
    return "local";
}

export const getBackendUrl = (url: string): string | undefined => {
    const env = getEnv(url);

    if (env === "prod") {
        return "https://consult.ai.cabinetoffice.gov.uk";
    } else if (env === "dev") {
        return "https://consult-dev.ai.cabinetoffice.gov.uk";
    } else if (env === "local") {
        return "http://localhost:8000";
    }
}

export const applyHighlight = (fullText: string, matchedText: string): string => {
    if (!matchedText) {
        return fullText;
    }
    const regex = new RegExp(matchedText, "gi");
    return fullText.replace(regex, match => `<span class="bg-yellow-300">${match}</span>`);
}