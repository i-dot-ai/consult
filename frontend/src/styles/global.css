@tailwind base;
@tailwind components;
@tailwind utilities;

:root {
    --color-primary: #c40678;
}

@view-transition {
    navigation: auto;
}

/* Temp until components ported to svelte */
iai-csv-download iai-silver-button button {
    display: flex;
    justify-content: center;
    align-items: center;
    gap: 0.5rem;
}
iai-csv-download iai-silver-button .material-symbols-outlined {
    font-size: 1.3rem;
}
table iai-progress-bar .container {
    max-width: 3rem;
}
table iai-progress-bar .container .bar {
    height: 0.5rem;
}
table iai-progress-bar .container .bar.full {
    border-radius: 0.5em;
}