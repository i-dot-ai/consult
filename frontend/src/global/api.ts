import type { APIContext } from "astro";
import path from "path";

import { getBackendUrl } from "./utils";

export const fetchBackendApi = async <T>(
  Astro: APIContext,
  endpoint: string,
  options: RequestInit | undefined = {},
): Promise<T> => {
  const url = path.join(getBackendUrl(), endpoint);
  const accessToken = Astro.cookies.get("access")?.value;

  const response = await fetch(url, {
    ...options,
    headers: {
      ...options.headers,
      contentType: "application/json",
      Authorization: `Bearer ${accessToken}`,
      cookie: Astro.request.headers.get("cookie") || "",
    },
  });

  if (!response.ok) {
    let errorBody;
    try {
      errorBody = await response.json();
    } catch {
      errorBody = await response.text();
    }
    throw { status: response.status, error: errorBody };
  }

  return response.json() as Promise<T>;
};
