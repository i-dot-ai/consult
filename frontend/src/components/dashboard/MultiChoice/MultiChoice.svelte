<script lang="ts">
    import clsx from "clsx";

    import { fade } from "svelte/transition";

    import { getPercentage } from "../../../global/utils";
    import { multiAnswerFilters } from "../../../global/state.svelte";

    import Progress from "../../Progress/Progress.svelte";
    import Panel from "../Panel/Panel.svelte";
    import TitleRow from "../TitleRow.svelte";
    import List from "../../svg/material/List.svelte";
    import Button from "../../inputs/Button/Button.svelte";


    export interface MultiChoiceAnswer {
        answer?: string;
        response_count: number;
    }
    interface Props {
        data: MultiChoiceAnswer[];
    }

    let {
        data = [],
    }: Props = $props();
</script>

<section class="my-4" transition:fade>
    <Panel border={true}>
        <TitleRow
            level={2}
            title="Multiple Choice Answers"
        >
            <List slot="icon" />
        </TitleRow>

        {@const total = data.map(item => item.response_count).reduce(
            (acc, curr) => acc + curr, 0,
        )}
        <Panel bg={true}>
            {#each data as item, i (i)}
                {@const percentage = getPercentage(item.response_count, total)}

                <div class="mb-1 last:mb-0">
                    <Button
                        variant="ghost"
                        size="xs"
                        fullWidth={true}
                        handleClick={() => {multiAnswerFilters.update(item.answer)}}
                        highlighted={multiAnswerFilters.applied(item.answer)}
                        highlightVariant="light"
                    >
                        <div class={clsx([
                            "flex",
                            "items-center",
                            "justify-between",
                            "gap-1",
                            "text-xs",
                            "mb-4",
                            "last:mb-0",
                            "w-full",
                            "p-1",
                        ])}>
                            <h3 class="w-1/2 text-left">
                                {item.answer}
                            </h3>

                            <div class={clsx([
                                "flex",
                                "flex-col",
                                "justify-center",
                                "items-center",
                                "gap-1",
                                "w-2/3",
                                "sm:flex-row",
                                "sm:w-1/2",
                                "sm:gap-4",
                            ])}>
                                <div class={clsx([
                                    "flex",
                                    "justify-between",
                                    "items-center",
                                    "w-full",
                                    "sm:justify-end",
                                ])}>
                                    <span>{percentage}%</span>

                                    <span class="sm:hidden">
                                        {item.response_count}
                                    </span>
                                </div>
                                <div class="w-full">
                                    <Progress value={percentage} />
                                </div>

                                <span class="hidden sm:block min-w-[4ch]">
                                    {item.response_count}
                                </span>
                            </div>
                        </div>
                    </Button>
                </div>
            {/each}
        </Panel>
    </Panel>
</section>