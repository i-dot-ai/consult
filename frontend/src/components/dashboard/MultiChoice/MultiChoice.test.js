import { afterEach, beforeEach, describe, expect, it } from "vitest";
import userEvent from "@testing-library/user-event";
import { render, cleanup, screen } from "@testing-library/svelte";

import { getPercentage } from "../../../global/utils";
import { multiAnswerFilters } from "../../../global/state.svelte";

import MultiChoice from "./MultiChoice.svelte";

let testData;

describe("MultiChoice", () => {
  beforeEach(() => {
    testData = {
      data: [
        {
          id: "id-1",
          text: "Multi Choice Answer 1",
          response_count: 10,
        },
        {
          id: "id-2",
          text: "Multi Choice Answer 2",
          response_count: 20,
        },
      ],
      totalCounts: 30,
    };
  });
  afterEach(() => {
    cleanup();
    multiAnswerFilters.reset();
  });

  it("should render data", () => {
    const { getByText, getAllByText } = render(MultiChoice, {
      data: testData.data,
    });

    testData.data.forEach((item) => {
      expect(getByText(item.text));
      expect(getAllByText(item.response_count).length).toBe(2);
      expect(
        getByText(
          `${getPercentage(item.response_count, testData.totalCounts)}%`,
        ),
      );
    });
  });

  it("should toggle multi answer filters when items clicked", async () => {
    const user = userEvent.setup();

    render(MultiChoice, { data: testData.data });

    // No filters applied yet
    expect(multiAnswerFilters.filters).toStrictEqual([]);

    const buttons = screen.getAllByRole("button");

    // First filter applied
    await user.click(buttons.at(0));
    expect(multiAnswerFilters.filters).toStrictEqual([testData.data[0].id]);

    // Second filter applied
    await user.click(buttons.at(1));
    expect(multiAnswerFilters.filters).toStrictEqual([
      testData.data[0].id,
      testData.data[1].id,
    ]);

    // First filter removed by clicking on it again
    await user.click(buttons.at(0));
    expect(multiAnswerFilters.filters).toStrictEqual([testData.data[1].id]);
  });
});
