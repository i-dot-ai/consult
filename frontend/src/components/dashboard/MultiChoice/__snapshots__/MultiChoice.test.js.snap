// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MultiChoice > should render data 1`] = `
<div>
  <section
    class="my-4"
  >
    <div
      class="h-full my-4 p-4 rounded-2xl border border-gray-200"
      data-testid="panel-component"
    >
      <!---->
      <div
        class="flex gap-2"
      >
        <div
          class="flex justify-center items-center"
        >
          <div
            class="bg-pink-50 p-2 rounded-md mt-1.5"
          >
            <svg
              class="fill-pink-700"
              height="1.3rem"
              role="img"
              viewBox="0 -960 960 960"
              width="1.3rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <!---->
              <path
                d="M280-600v-80h560v80H280Zm0 160v-80h560v80H280Zm0 160v-80h560v80H280ZM160-600q-17 0-28.5-11.5T120-640q0-17 11.5-28.5T160-680q17 0 28.5 11.5T200-640q0 17-11.5 28.5T160-600Zm0 160q-17 0-28.5-11.5T120-480q0-17 11.5-28.5T160-520q17 0 28.5 11.5T200-480q0 17-11.5 28.5T160-440Zm0 160q-17 0-28.5-11.5T120-320q0-17 11.5-28.5T160-360q17 0 28.5 11.5T200-320q0 17-11.5 28.5T160-280Z"
                xmlns="http://www.w3.org/2000/svg"
              />
              
              <!---->
            </svg>
            <!---->
          </div>
        </div>
        <!---->
         
        <div
          class="grow"
        >
          <!---->
          <h2
            class="text-neutral-700 text-lg"
            style=""
          >
            <!---->
            Multiple Choice Answers
            
            
          </h2>
          
          <!---->
           
          <!---->
        </div>
         
        <!---->
      </div>
      <!---->
       
      <div
        class="h-full my-4 p-4 rounded-2xl border border-gray-200 bg-neutral-50"
        data-testid="panel-component"
      >
        <!---->
        <div
          class="mb-1 last:mb-0"
        >
          <!---->
          <button
            aria-pressed="false"
            class="w-full text-xs cursor-pointer rounded-md py-0.5 px-1 border border-transparent transition-colors duration-300 self-start inline-flex gap-1 items-center hover:bg-gray-100 svelte-1vf7a7q"
            data-variant="ghost"
            role="button"
            tabindex="0"
          >
            <!---->
            <div
              class="flex items-center justify-between gap-1 text-xs mb-4 last:mb-0 w-full p-1"
            >
              <h3
                class="w-1/2 text-left"
              >
                Multi Choice Answer 1
              </h3>
               
              <div
                class="flex flex-col justify-center items-center gap-1 w-2/3 sm:flex-row sm:w-1/2 sm:gap-4"
              >
                <div
                  class="flex justify-between items-center w-full sm:justify-end"
                >
                  <span>
                    33.3%
                  </span>
                   
                  <span
                    class="sm:hidden"
                  >
                    10
                  </span>
                </div>
                 
                <div
                  class="w-full"
                >
                  <div
                    aria-valuemax="100"
                    aria-valuemin="0"
                    aria-valuenow="33.3"
                    class="relative h-2 w-full overflow-hidden rounded-[99999px] bg-neutral-300"
                    data-max="100"
                    data-melt-progress=""
                    data-state="loading"
                    data-value="33.3"
                    max="100"
                    role="meter"
                  >
                    <div
                      class="h-full w-full bg-primary transition-transform duration-[660ms] ease-[cubic-bezier(0.65,0,0.35,1)]"
                      style="transform: translateX(-66.7%);"
                    />
                  </div>
                  <!---->
                </div>
                 
                <span
                  class="hidden sm:block min-w-[4ch]"
                >
                  10
                </span>
              </div>
            </div>
            
            
          </button>
          
          <!---->
        </div>
        <div
          class="mb-1 last:mb-0"
        >
          <!---->
          <button
            aria-pressed="false"
            class="w-full text-xs cursor-pointer rounded-md py-0.5 px-1 border border-transparent transition-colors duration-300 self-start inline-flex gap-1 items-center hover:bg-gray-100 svelte-1vf7a7q"
            data-variant="ghost"
            role="button"
            tabindex="0"
          >
            <!---->
            <div
              class="flex items-center justify-between gap-1 text-xs mb-4 last:mb-0 w-full p-1"
            >
              <h3
                class="w-1/2 text-left"
              >
                Multi Choice Answer 2
              </h3>
               
              <div
                class="flex flex-col justify-center items-center gap-1 w-2/3 sm:flex-row sm:w-1/2 sm:gap-4"
              >
                <div
                  class="flex justify-between items-center w-full sm:justify-end"
                >
                  <span>
                    66.7%
                  </span>
                   
                  <span
                    class="sm:hidden"
                  >
                    20
                  </span>
                </div>
                 
                <div
                  class="w-full"
                >
                  <div
                    aria-valuemax="100"
                    aria-valuemin="0"
                    aria-valuenow="66.7"
                    class="relative h-2 w-full overflow-hidden rounded-[99999px] bg-neutral-300"
                    data-max="100"
                    data-melt-progress=""
                    data-state="loading"
                    data-value="66.7"
                    max="100"
                    role="meter"
                  >
                    <div
                      class="h-full w-full bg-primary transition-transform duration-[660ms] ease-[cubic-bezier(0.65,0,0.35,1)]"
                      style="transform: translateX(-33.3%);"
                    />
                  </div>
                  <!---->
                </div>
                 
                <span
                  class="hidden sm:block min-w-[4ch]"
                >
                  20
                </span>
              </div>
            </div>
            
            
          </button>
          
          <!---->
        </div>
        
        <!---->
      </div>
      <!---->
      <!---->
    </div>
    <!---->
  </section>
  
</div>
`;
